### ISSUE TITLE
LWS_CALLBACK_CLIENT_APPEND_HANDSHAKE_HEADER does not provide user context

### ISSUE THREAD

**Issue by smipi1** [warmcat/libwebsockets#509]
The `LWS_CALLBACK_CLIENT_APPEND_HANDSHAKE_HEADER` event does not set the `user` pointer to the value specified when calling `lws_client_connect_via_info()`.

Currently this is by design, but it limits proper OO integration of libwebsockets. This is especially problematic, where header information must be protocol or connection-specific.

**Comment by lws-team** [comment]
The logic for it being like that is that the ws connection protocol has not been negotiated at that time.

The client may send several protocols that it can support and the server will pick the one it thinks is "best" in the response.  So you can send "oldprot,myprot-v1,myprot-v2" and if the server supports myprot-v2, he might prefer that and assert that is the connection protocol in his response.

The per-connection "user" allocation would reflect whatever myprot-v2 needed then.  If something else, some other, perhaps radically different allocation might be used.  The point is that's not chosen until the server response is parsed, at least a network roundtrip after this callback.

Actually if you look in `struct lws_client_connect_info`, there is a `.userdata` member which lets you preallocate (with malloc) the user data for the client wsi if "you know better", lws will leave it alone then when the protocol gets decided.

Also, "proper" OO integration will need some care, you can't cleanly "proper" your way around lws basic features like needing to do things when the sink for the data is writable and needing LWS_PRE.

**Comment by smipi1** [comment]
My usecase is simple:
- I can establish trust in the server with the key + certificate installed on the server and setting `lws_context_creation_info::ssl_ca_filepath` to the appropriate root CA file.
- I want to establish trust in the client by adding authentication artifacts to the HTTP header.
- Because I manage connections to multiple servers, I allocate an application connection context.
- I set `lws_client_connect_info::userdata` to point to the application connection context.
- Whenever a callback fires, I am usually able to reconstruct the connection context from the `user` argument passed to `lws_callback_function`.
- The `user` argument however is set to `NULL` when `LWS_CALLBACK_CLIENT_APPEND_HANDSHAKE_HEADER` fires, even though I set it by passing a `pre-allocated lws_client_connect_info::userdata` to  `lws_client_connect_via_info()`
